api/http/management.go

```diff
@@ -371,23 +390,53 @@ Loop:
 				return err
 			}
 			if mr.Header.Proto == ws.ProtoTypeShell {
-				if mr.Header.MsgType == shell.MessageTypeShellCommand {
+				switch mr.Header.MsgType {
+				case shell.MessageTypeShellCommand:
+					timeNowUTC := int64(
+						float64(time.Now().UTC().UnixNano()) * 0.000001)
+					keystrokeDelay := timeNowUTC - lastKeystrokeAt
+					if keystrokeDelay >= keyStrokeDelayRecordingThresholdMs {
+						controlMsg := app.Control{
+							Type:           app.DelayMessage,
+							Offset:         recordedBytes,
+							DelayMs:        uint16(keystrokeDelay),
+							TerminalHeight: 0,
+							TerminalWidth:  0,
+						}
+						n, _ := controlRecorderBuffered.Write(
+							controlMsg.GetBytes())
+						l.Debugf("saving control delay message: %+v/%d",
+							controlMsg, n)
+						controlBytes += n
+						if controlBytes >= app.MessageSizeLimit {
+							l.Infof("closing session session_id=%s"+
+								"on control data limit reached.",
+								session.ID)
+							//see MEN-4448
+							break Loop
+						}
+					}
+					lastKeystrokeAt = timeNowUTC
 					b, e := recorderBuffered.Write(mr.Body)
 					if e != nil {
 						l.Errorf("session logging: "+
 							"recorderBuffered.Write(len=%d)=%d,%+v",
 							len(mr.Body), b, e)
 					}
 					recordedBytes += len(mr.Body)
+					session.BytesRecordedMutex.Lock()
```

Comment:
This mutex doesn't seem to do anything...

Reply:
it guards the BytesRecorded field which is updated here and read in the other go routine in `sendResizeMessage`

Reply:
But this doesn't prevent the data race; what if either go routine acquires the mutex twice?

Reply:
this one is under discussion in slack.